File: src/cpp/query_engine/StarNode.cc
#include "StarNode.h"
#include "LeadershipBroker.h"
#include "MessageBroker.h"

using namespace atom_space_node;
using namespace std;

// -------------------------------------------------------------------------------------------------
// Constructors and destructors

StarNode::StarNode(
    const string &node_id, 
    MessageBrokerType messaging_backend) :
    AtomSpaceNode(node_id, LeadershipBrokerType::SINGLE_MASTER_SERVER, messaging_backend) {

    this->is_server = true;
    this->join_network();
}

StarNode::StarNode(
    const string &node_id, 
    const string &server_id, 
    MessageBrokerType messaging_backend) : 
    AtomSpaceNode(node_id, LeadershipBrokerType::SINGLE_MASTER_SERVER, messaging_backend) {

    this->server_id = server_id;
    this->is_server = false;
    this->add_peer(server_id);
    this->join_network();
}

StarNode::~StarNode() {
}

// -------------------------------------------------------------------------------------------------
// DistributedAlgorithmNode virtual API

void StarNode::node_joined_network(const string &node_id) {
    if (this->is_server) {
        this->add_peer(node_id);
    }
}

string StarNode::cast_leadership_vote() {
    if (this->is_server) {
        return this->node_id();
    } else {
        return this->server_id;
    }
}

File: src/evolution/das_node/star_node.py
from hyperon_das_node import DistributedAlgorithmNode, LeadershipBrokerType, MessageBrokerType
import time


class StarNode(DistributedAlgorithmNode):
    is_server: bool
    server_id: str

    def __init__(
        self,
        node_id: str,
        server_id: str = None,
        messaging_backend: MessageBrokerType = MessageBrokerType.GRPC,
    ):
        # Call the parent constructor (DistributedAlgorithmNode)
        super().__init__(node_id, LeadershipBrokerType.SINGLE_MASTER_SERVER, messaging_backend)
        if server_id:
            # If server_id is provided, this is a client node
            self.server_id = server_id
            self.is_server = False
            self.add_peer(server_id)
        else:
            # If no server_id, this is a server node
            self.is_server = True

        # Join the network regardless of server/client
        self.join_network()
        time.sleep(1)

    def node_joined_network(self, node_id: str):
        if self.is_server:
            self.add_peer(node_id)

    def cast_leadership_vote(self) -> str:
        if self.is_server:
            return self.node_id()
        else:
            return self.server_id
